src/workerd/api/streams/standard.c++

```diff
@@ -3971,8 +3978,32 @@ jsg::Promise<void> TransformStreamDefaultController::write(
 jsg::Promise<void> TransformStreamDefaultController::abort(
     jsg::Lock& js,
     v8::Local<v8::Value> reason) {
-  error(js, reason);
-  return js.resolvedPromise();
+  KJ_IF_SOME(finish, algorithms.maybeFinish) {
+    return finish.whenResolved(js);
+  }
+  auto& promise = algorithms.maybeFinish.emplace(maybeRunAlgorithm(js, algorithms.cancel,
+      JSG_VISITABLE_LAMBDA(
+          (this, ref=JSG_THIS, reason = jsg::JsRef(js, jsg::JsValue(reason))),
+          (ref, reason),
```

Comment:
Just noticed that you moved this back to draft state, but LGTM overall. One thing I didn't understand were the JSG_VISITABLE_LAMBDA arguments here, even after looking at existing usage examples (why does the closure include two arguments here?)

Reply:
The `JSG_VISITABLE_LAMBA`  is a bit weird. The first set `(this, ref=JSG_THIS, reason = jsg::JsRef(...))` defines the captures for the lamba. The second set `(ref, reason)` defines the bits that are gc visited, and the third is the actual body of the lambda. The point is to ensure that `ref` and `reason` (in this case) are visible to the v8 gc visitation mechanism when the promise they are attached to here is gc visited. It's not the easiest thing to get your head around but it works.